### ISSUE TITLE
is ocnews-flag-star working?

### ISSUE THREAD

**Issue by aleprovencio** [newsboat/newsboat#743]
This is newsboat 2.18.0.

I have `ocnews-flag-star "s"` set on config, but when hitting "s" on an article `newsboat` tries to save it as a file instead of starring/flagging it. Behavior is the same when this option is not set.

**Comment by Minoru** [comment]
Hi! Newsboat has it own feature called "flags". Flags are one-letter marks than you can set on an article by pressing `^E`.

`ocnews-flag-star` specifies a flag which, when set on an article, will be translated into a star in ownCloud News.

So you open an article, press `^E`, type `s` and Enter—and now the article is starred in ownCloud News. You can remove the flag, and it will be un-starred.

And `s` simply saves the article, as you already found.

**Comment by aleprovencio** [comment]
Got it, thanks!
However, it does not seem to be syncing with nextcloud, are there any tests I can do?

**Comment by Minoru** [comment]
First of all, note that the flags are only synched when they change. So if you already changed them and *then* set `ocnews-flag-star`, the flag wasn't be synched to nextcloud.

If that doesn't solve it, you'll have to dive into code and add some logging to [OcNewsApi::update_article_flags](https://github.com/newsboat/newsboat/blob/a7490633b007fdea50144a94c7d419de0feb4b42/src/ocnewsapi.cpp#L171).

@dirb, you implemented ownCloud News API in Newsboat—are you still using? Can you check if stars are synchronized okay?

**Comment by aleprovencio** [comment]
Yeah, I was trying to sync a fresh flagged article on my tests, it really seem like an issue to me.

**Comment by dennisschagt** [comment]
I can reproduce this issue.

According to nextcloud-news documentation [1], marking a single item as starred can be done by HTTP `PUT` on route `/items/{feedId}/{guidHash}/star`.
It looks like Newsboat puts in the `guid` instead of the `guidHash`, resulting in an HTTP 405 error.

### guid and guidHash field in first sync
![image](https://user-images.githubusercontent.com/4629607/74604898-c5c32700-50c2-11ea-9f1c-65f5f45103c4.png)

### HTTP request/response pair for (failed) marking item as starred
![image](https://user-images.githubusercontent.com/4629607/74604911-eee3b780-50c2-11ea-9ee3-71485ddd3942.png)

[1] https://github.com/nextcloud/news/blob/cdd15c181875dbbc62844718365c47b099027b3a/docs/externalapi/Legacy.md#user-content-mark-an-item-as-starred

**Comment by Minoru** [comment]
Thanks! It looks like GUID itself isn't used in queries, so we can store guidHash in `rss_item.guid`, and guid in `rss_item.url` (because it's a link to the item).

**Comment by dennisschagt** [comment]
> It looks like GUID itself isn't used in queries, so we can store guidHash in `rss_item.guid`

Changing the `rss_item.guid` format will cause all items to be duplicated in the cache.
Some possible solutions/workarounds:
1. Create a script (or sqlite query) which users can execute to remove duplicates (only needs to be executed once)
    - Disadvantage: Requires effort from ocnews users, even those who are not interested in the "flag->starred" feature.
2. Remove duplicates in Newsboat itself
    - Disadvantage: I think this does not fit well into the current structure of Newsboat so it might add significant complexity.
3. Query the [`/items` endpoint](https://github.com/nextcloud/news/blob/master/docs/externalapi/Legacy.md#user-content-get-items) to map the item's `"id"` to its `"guidHash"` everytime we star/unstar an item.
    - Disadvantage: Wastes resources (bandwidth and server processing power) and takes more time due to extra request

I'm not happy with any of those solutions so please let me know if you have a better idea.

> and store guid in `rss_item.url` (because it's a link to the item).

Do you mean `rss_item.link`?
That is already filled with the value of the ocnews `"url"` field:
https://github.com/newsboat/newsboat/blob/9b289cd57bf8746bd665b17a1b06f798d6e47808/src/ocnewsapi.cpp#L242-L245

**Comment by Minoru** [comment]
> Changing the rss_item.guid format will cause all items to be duplicated in the cache.

Right. I didn't consider that.

> I'm not happy with any of those solutions so please let me know if you have a better idea.

Without re-reading their docs, the best I can suggest is taking option 3 and adding a cache to it: either a new column in `rss_item` table, or a new table altogether (using a foreign key on `rss_item.guid` or something like that).

> Do you mean rss_item.link?

I meant the `url` column in the `rss_item` table, but yeah, it maps to `rsspp::Item::link`.

> That is already filled with the value of the ocnews "url" field:

So much for that idea then :)
